src/ext/php7/curl_handlers.c

```diff
@@ -0,0 +1,612 @@
+            return SUCCESS;
+        }
+    }
+    zend_string_release(Z_STR(member));
+    return FAILURE;
+}
+
+// Should be run after request_init_hook is run *and* all the classes are loaded
+static bool _dd_load_curl_integration(void) {
+    if (!get_dd_trace_sandbox_enabled()) {
+        return false;
+    }
+
+    if (_dd_curl_integration_loaded) {
+        return true;
+    }
+
+    _dd_ArrayKVStore_ce = _dd_lookup_ce(ZEND_STRL("DDTrace\\Util\\ArrayKVStore"), 0);
+    _dd_GlobalTracer_ce = _dd_lookup_ce(ZEND_STRL("DDTrace\\GlobalTracer"), 0);
+    _dd_SpanContext_ce = _dd_lookup_ce(ZEND_STRL("DDTrace\\SpanContext"), 0);
+
+    if (!_dd_ArrayKVStore_ce || !_dd_GlobalTracer_ce || !_dd_SpanContext_ce) {
+        return false;
+    }
+
+    zend_string *format_headers = zend_string_init(ZEND_STRL("DDTrace\\Format::CURL_HTTP_HEADERS"), 0);
+    zval *format_headers_zv = zend_get_constant_ex(format_headers, NULL, ZEND_FETCH_CLASS_SILENT);
+    zend_string_release(format_headers);
+    if (format_headers_zv) {
+        ZVAL_COPY(&_dd_format_curl_http_headers, format_headers_zv);
+    } else {
+        return false;
+    }
+
+    zend_string *curlopt_httpheader = zend_string_init(ZEND_STRL("CURLOPT_HTTPHEADER"), 0);
+    zval *curlopt_httpheader_zv = zend_get_constant_ex(curlopt_httpheader, NULL, ZEND_FETCH_CLASS_SILENT);
+    zend_string_release(curlopt_httpheader);
+    if (curlopt_httpheader_zv) {
+        ZVAL_COPY(&_dd_curl_httpheaders, curlopt_httpheader_zv);
+    } else {
+        return false;
+    }
+
+    _dd_curl_integration_loaded = true;
+    return true;
+}
+
+static zval *_dd_ArrayKVStore_getForResource(zval *ch, zval *format, zval *value, zval *retval) {
+    zval args[3] = {*ch, *format, *value};
+
+    int result = ddtrace_call_method(NULL, _dd_ArrayKVStore_ce, &_dd_ArrayKVStore_getForResource_fe,
+                                     ZEND_STRL("getForResource"), retval, 3, args);
+
+    return (result == SUCCESS) ? retval : NULL;
+}
+
+static void _dd_ArrayKVStore_putForResource(zval *ch, zval *format, zval *value) {
+    zval args[3] = {*ch, *format, *value};
+    zval retval;
+    ZVAL_UNDEF(&retval);
+
+    ddtrace_call_method(NULL, _dd_ArrayKVStore_ce, &_dd_ArrayKVStore_putForResource_fe, ZEND_STRL("putForResource"),
+                        &retval, 3, args);
+
+    zval_dtor(&retval);
+}
+
+/* curl_close {{{ */
+static void (*_dd_curl_close_handler)(INTERNAL_FUNCTION_PARAMETERS) = NULL;
+
+ZEND_FUNCTION(ddtrace_curl_close) {
+    zval *ch;
+
+    if (_dd_load_curl_integration() &&
+        zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS(), "r", &ch) == SUCCESS) {
+        zval retval;
+        zend_call_method_with_1_params(NULL, _dd_ArrayKVStore_ce, &_dd_ArrayKVStore_deleteResource_fe, "deleteresource",
+                                       &retval, ch);
+    }
+
+    _dd_curl_close_handler(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+}
+
+static void _dd_instrument_curl_close(void) {
```

Comment:
I'm a bit confused by this function name as it mentions instrumentation but I don't see any instrumentation logic. I'm guessing this is just a hook for the custom handler?

Same goes for `_dd_instrument_curl_copy_handle()`, `_dd_instrument_curl_exec()`, `_dd_instrument_curl_setopt()`, and `_dd_instrument_curl_setopt_array()` below. :)

Reply:
This series of functions overrides the internal function handlers -- I could change "instrument" to "install"?